website/server/libs/tasks/index.js

```diff
@@ -232,62 +194,88 @@ export async function getTasks (req, res, options = {}) {
-      }
+    order = owner.tasksOrder[type];
+  } else if (!type) {
+    Object.values(owner.tasksOrder).forEach(taskOrder => {
+      order = order.concat(taskOrder);
     });
-
-    // Remove empty values from the array and add any unordered task
-    orderedTasks = _.compact(orderedTasks).concat(unorderedTasks);
-    return orderedTasks;
+  } else {
+    return tasks;
   }
-  return tasks;
+
+  let orderedTasks = new Array(tasks.length);
+  const unorderedTasks = []; // what we want to add later
+
+  tasks.forEach((task, index) => {
+    const taskId = task._id;
+    const i = order[index] === taskId ? index : order.indexOf(taskId);
+    if (i === -1) {
+      unorderedTasks.push(task);
+    } else {
+      orderedTasks[i] = task;
+    }
+  });
+
+  // Remove empty values from the array and add any unordered task
+  orderedTasks = _.compact(orderedTasks).concat(unorderedTasks);
+  return orderedTasks;
 }
 
-// Takes a Task document and return a plain object of attributes that can be synced to the user
-export function syncableAttrs (task) {
-  const t = task.toObject(); // lodash doesn't seem to like _.omit on Document
-  // only sync/compare important attrs
-  const omitAttrs = ['_id', 'userId', 'challenge', 'history', 'tags', 'completed', 'streak', 'notes', 'updatedAt', 'createdAt', 'group', 'checklist', 'attribute'];
-  if (t.type !== 'reward') omitAttrs.push('value');
-  return _.omit(t, omitAttrs);
+
+function canNotEditTasks (group, user, assignedUserId) {
+  const isNotGroupLeader = group.leader !== user._id;
+  const isManager = Boolean(group.managers[user._id]);
+  const userIsAssigningToSelf = Boolean(assignedUserId && user._id === assignedUserId);
+  return isNotGroupLeader && !isManager && !userIsAssigningToSelf;
 }
 
-/**
- * Moves a task to a specified position.
- *
- * @param  order  The list of ordered tasks
- * @param  taskId  The Task._id of the task to move
- * @param  to A integer specifying the index to move the task to
- *
- * @return Empty
- */
-export function moveTask (order, taskId, to) {
-  const currentIndex = order.indexOf(taskId);
-
-  // If for some reason the task isn't ordered (should never happen), push it in the new position
-  // if the task is moved to a non existing position
-  // or if the task is moved to position -1 (push to bottom)
-  // -> push task at end of list
-  if (!order[to] && to !== -1) {
-    order.push(taskId);
-    return;
+
+async function getGroupFromTaskAndUser (task, user) {
+  if (task.group.id && !task.userId) {
+    const fields = requiredGroupFields.concat(' managers');
+    return Group.getGroup({ user, groupId: task.group.id, fields });
   }
+  return null;
+}
 
-  if (currentIndex !== -1) order.splice(currentIndex, 1);
-  if (to === -1) {
-    order.push(taskId);
-  } else {
-    order.splice(to, 0, taskId);
+
+async function getChallengeFromTask (task) {
+  if (task.challenge.id && !task.userId) {
+    return Challenge.findOne({ _id: task.challenge.id });
```

Comment:
.exec() at the end here, it's confusing but findOne doesn't return a real promise untill you call .exec()

Reply:
Nice catch!